### ISSUE TITLE
Extremely slow execution of select query that has `with` in statement

### ISSUE THREAD

**Issue by Paper-Folding** [xerial/sqlite-jdbc#1015]
### **Describe the bug**
When I use select query which has `with` statement, the query gets really slow. For example I use this sql in my reproduce example below:  
```sql
with parent as (
    with parent_feature as (
        select id, min(sub_id) as sub_id from sample group by id
    )
    select p.* from parent_feature as pf left join sample as p on pf.id = p.id and pf.sub_id = p.sub_id
)
select a.*, b.id as child_id, b.sub_id as child_sub_id
from parent as a left join sample as b
on a.id = b.id and a.sub_id != b.sub_id
order by a.id, a.sub_id;
```
*Note: I use `with` twice to join the table again, which seems redundant here, but it was because in my database under production, there are other fields I need to yield, so that's why I keep the redundancy here.

### **To Reproduce**
Please check out this repo I created just now for reproduction: https://github.com/Paper-Folding/sqlite-jdbc-slow-query-reproduce

1. Here is the structure of my sample database, it has only one single table named "sample":  
```sql
create table if not exists sample(
    id integer not null,
    sub_id integer default 0 not null,
);
```
The table has this idea: for each record, the` id` and `sub_id` are both not unique; if two records have a same `id`, they must have different `sub_id`.  

2. Now execute the sql mentioned above for this table, here is the full java code that does that:
```java
package org.example;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.Statement;

public class Main {
    public static void main(String[] args) {

        try {
            Class.forName("org.sqlite.JDBC");
            Connection c = DriverManager.getConnection("jdbc:sqlite:sample.db");
            Statement s = c.createStatement();
            String sql = """
                        with parent as (
                            with parent_feature as (
                                select id, min(sub_id) as sub_id from sample group by id
                            )
                            select p.* from parent_feature as pf left join sample as p on pf.id = p.id and pf.sub_id = p.sub_id
                        )
                        select a.*, b.id as child_id, b.sub_id as child_sub_id
                        from parent as a left join sample as b
                        on a.id = b.id and a.sub_id != b.sub_id
                        order by a.id, a.sub_id;
                    """;

            System.out.println("start querying");
            ResultSet rs = s.executeQuery(sql);
            System.out.println("finish querying");
            while (rs.next()) {
                System.out.println(rs.getInt("id"));
                System.out.println(rs.getInt("sub_id"));
            }
            s.close();
            c.close();
        } catch (Exception ignored) {

        }
    }
}
```

3. You should see that the execution is quite slow from "start querying" to "finish querying" in console (about 20 seconds on my machine).

### **Expected behavior**
The execution should be really fast within 1 second, for example, If I use sqlite db browser to execute the sql statement or python's `sqlite3` lib, they yield result super fast.

### **Environment:**
 - OS: Windows 11 x64
 - CPU architecture: x64 Intel i9-13900HX
 - sqlite-jdbc version 3.43.2.2

**Additional context**
If I add `limit 1` to the end of the sql I wrote above, it does not speed up even though it yields only one record!  
If I use old version `3.14.2.1` metioned in issue #160, it gets fast, so I really doubt some changes made causing the speed drop, please investigate, thank you.

**Comment by Paper-Folding** [comment]
Oh, and the database file I mentioned above is in reproduce repo, too, see here: https://github.com/Paper-Folding/sqlite-jdbc-slow-query-reproduce/blob/master/sample.db

**Comment by gotson** [comment]
This seems related to sqlite itself more than the driver.

**Comment by Paper-Folding** [comment]
> This seems related to sqlite itself more than the driver.

Really? It executes super fast in python `sqlite3` lib!

**Comment by Paper-Folding** [comment]
> This seems related to sqlite itself more than the driver.

I think you are right, python also got really slow after upgrading python to latest version. And as I test with different versions of sqlite-jdbc, I found 3.40.1.0 is normal and 3.41.0.0 is really slow, https://github.com/xerial/sqlite-jdbc/compare/3.40.1.0...3.41.0.0, this problem seems to happen at that time....

**Comment by Paper-Folding** [comment]
Hmmmm.... Looks like the issue was fixed in 3.44.0.0....
